src/contrib/cluster/Akka.Cluster.Tools/Singleton/OldestChangedBufferState.cs

```diff
@@ -0,0 +1,74 @@
+﻿// -----------------------------------------------------------------------
+//  <copyright file="OldestChangedBufferState.cs" company="Akka.NET Project">
+//      Copyright (C) 2009-2024 Lightbend Inc. <http://www.lightbend.com>
+//      Copyright (C) 2013-2024 .NET Foundation <https://github.com/akkadotnet/akka.net>
+//  </copyright>
+// -----------------------------------------------------------------------
+
+#nullable enable
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace Akka.Cluster.Tools.Singleton;
+
+/// <summary>
+/// Immutable data object that represents the state of the oldest changed buffer.
+/// </summary>
+internal sealed record OldestChangedBufferState
```

Comment:
New immutable data structure designed to do two things:

1. Predict who the next leader should be
2. Not actually allow the leader to be changed until the previous leader has exited the cluster <-- THIS IS THE KEY BUGFIX.